Order Creation: Present New Order flow as a modal

itsmeichigo opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of #6381
⚠️ Please make sure that this PR is merged before #6387 ⚠️ 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
When working on the split view for the Orders tab in PR #6387, I found that presenting the New Order flow using the navigation stack doesn't look good on tablets. After discussing with the team (internal ref: p1646791790098809-slack-CGPNUU63E), it's decided that we should present the flow modally and disable swipe-to-dismiss to prevent accidental dismissal and loss of data.

Solution:
- Add a cancel button on the New Order button to handle manual dismissal
- Update the presentation of the New Order flow to use a non-dismissible modal.
- Also: update the presentation of Simple Payment flow to be non-dismissible too.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
- Log in to a valid account
- Enable Order Creation in Settings > Experimental Features > Order Creation
- Select Orders tab > tap the (+) button and select Create new order
- Notice that the New Order screen is presented modally and cannot be dismissed by swiping down.
- Tap the Cancel button to dismiss the New Order screen. Notice that the screen is dismissed successfully.
- Tap the (+) button on the Orders screen and select Simple Payment. Notice the same behavior on the Simple Payment screen.

### Screenshots
<!-- Include before and after images or gifs when appropriate. -->
https://user-images.githubusercontent.com/5533851/157370382-73078657-ccea-4fe5-aca8-fc9422720180.mp4

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

itsmeichigo committed:
Add cancel button to New Order screen

itsmeichigo committed:
Present new order flow modally

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -10,7 +10,12 @@ final class NewOrderHostingController: UIHostingController<NewOrder> {
+        super.init(rootView: view)
+    }
+
+    func setDismissAction(_ dismissAction: @escaping () -> Void) {
+        rootView.dismissAction = dismissAction
     }
```
We could simplify this by setting the dismiss action in the `init` method, rather than having a separate function for it, similar to how we do for the `SimplePaymentsAmount` view:

https://github.com/woocommerce/woocommerce-ios/blob/c04248d75b06b25c318ba4e60443521ecb1e259b/WooCommerce/Classes/ViewRelated/Orders/Simple%20Payments/Amount/SimplePaymentsAmount.swift#L31-L34

itsmeichigo replied:
Nice suggestion! Updated in d6dabf46b46ddbc1b6d61b06b37700416882b86a 👍

rachelmcr reviewed (commented):
Works great! I have a question about the New Order behavior, though. Now that New Order is presented as a modal, do you think we should change the behavior of the sub-screens in that flow so they aren't presented as additional modals on top of this view?

Here's how it works now when you open a screen to add something to the new order:

https://user-images.githubusercontent.com/8658164/157445350-04eb73a2-5ac9-4f4f-b7db-d22f14dd15b2.mp4

Ecarrion reviewed (commented):
Hi @itsmeichigo, thanks for working on this.

I rebased the split-view branch on this one and noticed a few problems. I think we should be addressed them!

## SImple Payments Cancel

With this change, the Simple Payments flow can't be canceled. 

https://user-images.githubusercontent.com/562080/157453311-9bf9d66c-cbf7-4310-8b80-a8b0188b7890.MP4

I think we have to:

- Show an alert when trying to dismiss the modal. This can be done by using the `UIAdaptivePresentationControllerDelegate` on `SimplePaymentsAmountHostingController`
or
- Provide a cancel button on each screen(which is not a very good UX in my opinion)

## New Order Creation

After creating a new order, the order detail screen is pushed on the first navigation controller of the split view.

https://user-images.githubusercontent.com/562080/157453894-0d0f502b-7e87-419c-bff0-e98a8f09875c.MP4

## New Order Animations

I'm not very sure why, but with this change, dismiss animations from the new order screen are broken.


https://user-images.githubusercontent.com/562080/157454090-36eb7392-7244-48e1-80b0-3c33aea08534.mov

itsmeichigo committed:
Simplify dismiss logic for New Order screen

itsmeichigo committed:
Add alert when trying to dismiss the simple payment flow

itsmeichigo committed:
Fix line length violations

itsmeichigo committed:
Show alert when swiping to dismiss New Order screen

itsmeichigo committed:
Remove cancel button on New Order screen

itsmeichigo commented:
Thank you @rachelmcr and @Ecarrion for the reviews! 

I have added a few updates to present the dismiss alert when swiping to dismiss the Simple Payments and New Order flows. What I'm concerned about is when to allow swiping to dismiss in Simple Payment - I'm currently enabling that in the last step (please look at the video below). Let me know if this needs to be changed.

https://user-images.githubusercontent.com/5533851/157591051-45596df5-28bf-4d68-851b-9ae2083ad90d.mp4

> Now that New Order is presented as a modal, do you think we should change the behavior of the sub-screens in that flow so they aren't presented as additional modals on top of this view?

I think this would be nice, but it'll be a big change and needs to be handled in a separate PR. Also, I'm not sure if we should do that, so tagging @woocommerce/mobile-designers for opinions.

> After creating a new order, the order detail screen is pushed on the first navigation controller of the split view.

Thank Ernesto for finding this! I'll make sure to fix it in my split view PR 👍 

> I'm not very sure why, but with this change, dismiss animations from the new order screen are broken.

This is weird, probably because stacking modals isn't recommended. We should probably switch to a navigation stack for the sub-screens as Rachel suggested.

This PR is ready for another look!

itsmeichigo committed:
Display action sheet instead of alert

itsmeichigo committed:
Check for changes made in new order

itsmeichigo committed:
Switch to displaying actionsheet for confirmation

itsmeichigo committed:
Add unit tests for checking changes in new order view model

itsmeichigo committed:
Fix check for disabling swipe to dismiss on simple payment

itsmeichigo commented:
Update: following the discussion on the Slack thread p1646791790098809-slack-CGPNUU63E, I've updated to show action sheet instead of the confirm alert as in the last comment. I also added checks to only disable swipe-to-dismiss gesture when there's any change made. Here's a demo:

https://user-images.githubusercontent.com/5533851/157622374-9c306432-eaff-44cf-9875-0d54834b074e.mp4

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -8,20 +8,81 @@ final class NewOrderHostingController: UIHostingController<NewOrder> {
+        guard viewModel.hasChanges else {
+            return true
+        }
+        let actionSheet = UIAlertController(title: Localization.dismissAlertTitle, message: nil, preferredStyle: .actionSheet)
```
What do you think of using the `presentDiscardChangesActionSheet` helper from `UIAlertController+Helpers` here? The strings are a little different but I think they're ok in this context, and it would avoid duplicating the alert setup (here and on the Simple Payments screen).

itsmeichigo replied:
#TIL - thanks Rachel! Updated this in my latest changes.

rachelmcr commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrderViewModel.swift:
```diff
@@ -12,6 +12,16 @@ final class NewOrderViewModel: ObservableObject {
 
+    /// Indicates whether user has made any changes
+    ///
+    var hasChanges: Bool {
+        productRows.isNotEmpty ||
+        selectedProductViewModel != nil ||
+        currentOrderStatus != .pending ||
+        paymentDataViewModel.hasChangesMade ||
+        customerDataViewModel.isDataAvailable
+    }
```
We should be able to simplify this to check that the order isn't an empty new order:

```suggestion
    var hasChanges: Bool {
        orderSynchronizer.order != OrderFactory.emptyNewOrder
    }
```
(In which case we won't need the new `paymentDataViewModel.hasChangesMade` property.)

itsmeichigo replied:
Thanks for letting me know about this too! This is so much nicer 💯

rachelmcr reviewed (commented):
Thanks for adding the action sheet! I have a couple suggestions that could simplify the action sheet setup and checking for pending changes on the new order screen.

Ecarrion commented:
Hi @itsmeichigo!

Regarding the multiple modal views, does apple says it is not recommended? I think it is an expected behavior, the mail app uses it for example.

https://user-images.githubusercontent.com/562080/157681014-f61b034c-beed-4f44-8dd7-54ccd8cd6bdf.MP4

Also, I think that we may be able to do something with the missing dismiss animation, because presenting the "Customer Details Screen" does have a dismiss animation.

https://user-images.githubusercontent.com/562080/157681323-d6a5bf38-5aeb-4a3e-9b56-a33b0304db7a.mov

Regarding:

> What I'm concerned about is when to allow swiping to dismiss in Simple Payment - I'm currently enabling that in the last step (please look at the video below). Let me know if this needs to be changed.

I think we should show the alert/sheet on the last screen too! Once we add the auto-draft support for it the user won't be able to go back to that order unless a payment method is selected and the order status is changed!

---

**Lastly, but most important** is that @rachelmcr and I are a bit concerned about introducing navigation changes(and potential regressions) when we are about to finish & ship the Order Creation M2. 😟 

One thing that could help would be to split "Order Creation" changes and "Simple Payments" changes into separate PRs so we can test them more thoroughly and in isolation?

Another alternative would be to add them behind a feature flag for extra safety. I'm not sure if that is set up or you plan to merge everything in one go.

I think a local-feature flag could be useful as we don't know what other edge cases we are not handling.

What do you think?

itsmeichigo committed:
Use simpler logic to check for changes in new order

itsmeichigo committed:
Use presentDiscardChangesActionSheet in New Order

itsmeichigo committed:
Use presentDiscardChangesActionSheet in Simple Payment

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-order-creation-presentation

itsmeichigo committed:
Only present New Order modally when order split view is enabled

itsmeichigo committed:
Hide updates in New Order under feature flag for orders split view

itsmeichigo committed:
Remove hasChangesMade in PaymentDataViewModel

itsmeichigo committed:
Revert changes to SimplePaymentAmount when order split view is disabled

itsmeichigo committed:
Only set delegate for the presentationController when order split view is enabled

itsmeichigo committed:
Remove dismissAction from New Order

itsmeichigo committed:
Remove localization from SimplePaymentsAmountHostingController

itsmeichigo commented:
Thanks, @rachelmcr and @Ecarrion for reviewing again. I really appreciate your input.

> Regarding the multiple modal views, does apple says it is not recommended? I think it is an expected behavior, the mail app uses it for example.

My apologies, I was bold to state that so let me take it back. The loss of animation is really an interesting case, my wild guess is that it has something to do with SwiftUI. But I'm not going to state anything again because I clearly don't know anything here (don't shoot me).

I'll try to figure out the reason afterward if it's alright because I'm hiding these changes under the feature flag for the orders split view. I'll add it to my checklist to make sure it doesn't slip away before releasing the feature. But please feel free to give some insight if you have some clue because that'll help a lot!

> I think we should show the alert/sheet on the last screen too! Once we add the auto-draft support for it the user won't be able to go back to that order unless a payment method is selected and the order status is changed!

Thanks for confirming! I updated this in my latest changes.

Again, my apologies to the team for underestimating the effects of my changes. I've hidden all these changes under a feature flag for the orders split view project. Considering the size of the changes, I'm keeping the changes to Simple Payments and New Order in this PR, but let me know if you strongly believe that I should split them.

I hope my changes look better now, so please take another look and don't be mad at me 😅

itsmeichigo committed:
Remove redundant constant assignment for new order view

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/NewOrder.swift:
```diff
@@ -8,14 +8,44 @@ final class NewOrderHostingController: UIHostingController<NewOrder> {
+        guard viewModel.hasChanges else {
+            return true
+        }
+        UIAlertController.presentDiscardChangesActionSheet(viewController: self, onDiscard: { [weak self] in
+            self?.dismiss(animated: true, completion: nil)
+        })
```
For a slightly better experience, this bit can go in the `presentationControllerDidAttemptToDismiss` which is called after this method return `false` and gives a nice bounce feedback to the user.

itsmeichigo replied:
Good idea! Updated in 2ba3f08901899bc1d77aa0e8181273c3ad564cdc.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Amount/SimplePaymentsAmount.swift:
```diff
@@ -77,7 +79,19 @@ extension SimplePaymentsAmountHostingController: UIAdaptivePresentationControlle
+            return true
+        }
+
+        UIAlertController.presentDiscardChangesActionSheet(viewController: self, onDiscard: { [weak self] in
+            self?.dismiss(animated: true, completion: nil)
+        })
```
Same comment as as in `NewOrderHostingController` but making sure we take into account `rootView.viewModel.disableViewActions`. to not present the sheet while a network request is being made.

itsmeichigo replied:
Updated in 2f58a13fb04fa6125a81b04ae22647aaf9e34ef6.

Ecarrion replied:
I think this needs also to be under the feature flag right? 

If not, in production, this alert would be presented only if the user tries to dismiss the view while a network request is being made. But it will not be presented in a regular dismissal, which would be a bit inconsistent.

itsmeichigo replied:
Good point - updated in 7392ae1fcc40345d5d169708e45c248416ccb5dd.

Ecarrion commented on WooCommerce/Classes/ViewRelated/Orders/Simple Payments/Amount/SimplePaymentsAmount.swift:
```diff
@@ -77,7 +79,19 @@ extension SimplePaymentsAmountHostingController: UIAdaptivePresentationControlle
     func presentationControllerShouldDismiss(_ presentationController: UIPresentationController) -> Bool {
-        !rootView.viewModel.disableViewActions
+        guard ServiceLocator.featureFlagService.isFeatureFlagEnabled(.splitViewInOrdersTab) else {
+            return !rootView.viewModel.disableViewActions
```
This condition is needed to prevent the view to be dismissed (or the cancel confirmation to be shown) while a network request is being made.

So I think we need it even if `. splitViewInOrdersTab` is enabled.

itsmeichigo replied:
I think this might be redundant since we're already checking if there's an amount entered - and it's highly likely that we only have a network request made with an amount. But I updated the check for `shouldEnableSwipeToDismiss` in b6469212b5854a5c7c045e1bae3235a70bb139ef for extra safety.

Ecarrion reviewed (changes requested):
Thanks for adding the feature flag @itsmeichigo, changes look much better but I think we still need some few adjustments!

itsmeichigo committed:
New Order: only present discard alert on presentationControllerDidAttemptToDismiss

itsmeichigo committed:
Simple Payments: present discard alert in presentationControllerDidAttemptToDismiss

itsmeichigo committed:
Simple payments: Only enable swiping to dismiss if the view is not loading

itsmeichigo commented:
Updated the changes with your suggestions @Ecarrion, please take another look thanks.

itsmeichigo committed:
Merge branch 'trunk' into issue/6381-order-creation-presentation

itsmeichigo committed:
Only display discard alert on simple payments when split view is enabled

Ecarrion reviewed (approved):
Thanks for bearing with me @itsmeichigo 😇 !
This looks great! 🚀
